### DISCUSSION TITLE
How does `commandInHtmlFormat` setting work?

### DISCUSSION THREAD

**Discussion by kurnevsky** [scalameta/metals#4146] [Q&A] [answered]
I'm trying to implement links handling in metals doctor and stacktrace analizer for emacs. It's not clear for me how this setting work (there is not much documentation regarding it either). I see that if I set it to `vscode` I get links in a format
```
"command:metals.metals-goto-location?[{\"uri\":\"metalsDecode:file%3A%2F%2F%2Fprojectname%2Fprojectname.metals-buildtarget\",\"range\":{\"start\":{\"line\":0,\"character\":0},\"end\":{\"line\":0,\"character\":0}},\"otherWindow\":false}]"
```

1. Is it a [client command](https://scalameta.org/metals/docs/integrations/new-editor#metals-client-commands) that's handled by vscode plugin? Or it's somehow passed to metals and it interprets this command?
2. If it's a client command then how these `metalsDecode:` urls are handled? Are there other types of urls?
3. I see that there is a command `command:metals.reset-choice?[\"Build+import\"]` and in case it's a client command it's not documented either. Are there other commands that can be present in html?
4. Do all of these commands have `metals.` prefix in links? Usually commands are not prefixed this way.

It would be nice if this was documented :)

Added:
Found https://scalameta.org/metals/docs/integrations/new-editor#decode-file - is it what vscode plugin uses when handles these links?

**Comment by tgodzik** [comment, marked as answer]
> Is it a [client command](https://scalameta.org/metals/docs/integrations/new-editor#metals-client-commands) that's handled by vscode plugin? Or it's somehow passed to metals and it interprets this command?

Visual Studio Code allows for the commands to be encoded as links, which means that when user clicks on them they will invoke that commands.

> If it's a client command then how these metalsDecode: urls are handled? Are there other types of urls?

There are explained in  https://scalameta.org/metals/docs/integrations/new-editor/#decode-file it's just a command with an URI

> I see that there is a command command:metals.reset-choice?[\"Build+import\"] and in case it's a client command it's not documented either. Are there other commands that can be present in html?

In VS Code all the commands can be shown as links, but not sure if we can do something like that for emacs? We already do a bit different format for sublime, but this is very niche thing that we might need to rework. For decorations we will want for sure to use the new decoration LSP methods, which would make it much better with navigation and would mean the links in most places are not needed.

> Do all of these commands have metals. prefix in links? Usually commands are not prefixed this way

They can be with or without the prefix, prefix might be needed if there are other similar commands provided by other LSP extensions.

**Reply by kurnevsky** [reply]
Thanks.

> In VS Code all the commands can be shown as links, but not sure if we can do something like that for emacs? We already do a bit different format for sublime, but this is very niche thing that we might need to rework. For decorations we will want for sure to use the new decoration LSP methods, which would make it much better with navigation and would mean the links in most places are not needed.

The question was about client command - there is a list of client commands in the [docs](https://scalameta.org/metals/docs/integrations/new-editor/#metals-client-commands) but this particular command is not mentioned there (it's mentioned only as server command which is a different thing). So I wonder what other command can be there.

I used vscode format - emacs can handle it: https://github.com/emacs-lsp/lsp-metals/pull/77 So now if it changes don't forget about emacs :)

**Reply by tgodzik** [reply]
>The question was about client command - there is a list of client commands in the [docs](https://scalameta.org/metals/docs/integrations/new-editor/#metals-client-commands) but this particular command is not mentioned there (it's mentioned only as server command which is a different thing). So I wonder what other command can be there.

That should be https://scalameta.org/metals/docs/integrations/new-editor/#goto-location I think? Or do you mean another command.

> I used vscode format - emacs can handle it: https://github.com/emacs-lsp/lsp-metals/pull/77 So now if it changes don't forget about emacs :)

Sure!

**Reply by kurnevsky** [reply]
No, I mean `metals.reset-choice` - it can be present in the doctor html, and it seems it's a client command.

**Reply by tgodzik** [reply]
Ach, there seems to be a bit of a mess there, yeah. Sorry about that :disappointed: 

Raised a new discussion in https://github.com/scalameta/metals/issues/4160

**Reply by kpodsiad** [reply]
Regarding `metals.reset-choice` and other, similar commands. There are few commands which can be described as "client has to proxy this command to server".
